### ISSUE TITLE
Unexpected reordering of memory allocation when using GC.GetAllocatedBytesForCurrentThread() when running optimized

### ISSUE THREAD

**Issue by madelson** [dotnet/runtime#96836]
### Description

I'm using `GC.GetAllocatedBytesForCurrentThread()` in a unit test to verify zero-allocation behavior. My test is like this:

```
var before = GC.GetAllocatedBytesForCurrentThread();
// do zero-allocation thing
var delta = GC.GetAllocatedBytesForCurrentThread() - before;
Assert.That(delta, Is.EqualTo(0));
```

I found that this was failing in release mode, where `delta` as 24 rather than zero. This happened even when I commented out all lines between the 2 GC checks.

### Reproduction Steps

Here is a simple program which repros the issue for me in release mode but not in debug mode:

```
	var before = GC.GetAllocatedBytesForCurrentThread();
	var delta = GC.GetAllocatedBytesForCurrentThread() - before;
	//Thread.MemoryBarrier(); // uncommenting this fixes the problem
	Console.WriteLine((object)delta); // 0 in debug mode, 24 in release mode
```

### Expected behavior

I would expect to always get 0

### Actual behavior

I get 24 in release mode

### Regression?

Not sure

### Known Workarounds

Thread.MemoryBarrier seems to work.

### Configuration

.NET 8, Windows 11, x64

**Comment by jkotas** [comment]
The JIT is reordering the allocation of the `Int32` box to be in front of `GetAllocatedBytesForCurrentThread` call as a side-effect of various optimizations. @dotnet/jit-contrib This reordering does not seem to be profitable. Do 

However, small allocations are considered to be side-effect free and the JIT can reorder or remove them as it sees a fit in general. Issues like this are expected if you try to use GetAllocatedBytesForCurrentThread. Also, allocations can happen as side-effects of JITing.

```
       call     System.GC:GetAllocatedBytesForCurrentThread():long
       mov      rbx, rax
       mov      rcx, 0x7FFF6F6894F0
       call     CORINFO_HELP_NEWSFAST
       mov      rsi, rax
       call     System.GC:GetAllocatedBytesForCurrentThread():long
       sub      rax, rbx
       mov      qword ptr [rsi+0x08], rax
       mov      rcx, rsi
       call     [System.Console:WriteLine(System.Object)]
```

**Comment by EgorBo** [comment]
Minimal repro:
```cs
    static void Foo()
    {
        long before = GC.GetAllocatedBytesForCurrentThread();
        long delta = GC.GetAllocatedBytesForCurrentThread() - before;
        Console.WriteLine((object)delta);
    }
```

Looks like `impImportAndPushBox` could spill the trees first to avoid this behaviour? Simply changing this
https://github.com/dotnet/runtime/blob/f3750be98ec3dadc88bd65d908882f0a50c1bedc/src/coreclr/jit/importer.cpp#L3258
from NONE to SPILL_ALL fixes the issue, but I presume it might ruin some de-box opts?

**Comment by EgorBo** [comment]
So my initial conclusion that this behavior helps us to optimize boxings in certain cases (e.g. from GDVs?)
it can be "fixed" like this: https://github.com/EgorBo/runtime-1/commit/bbf583a1d94d7d7c68d1d73fc6bff2f8757d9e1c (fixes the issue) but brings a lot of size regressions. Since, as Jan noted, allocation is side-effect free (we ignore OOM as a side-effect) it's legal to keep it as is... we can intrinsify the GC call if needed.

cc @dotnet/jit-contrib

**Comment by jkotas** [comment]
> we can intrinsify the GC call if needed.

What do you mean?

**Comment by EgorBo** [comment]
> > we can intrinsify the GC call if needed.
> 
> What do you mean?

To basically always spill `GC.GetAllocatedBytesForCurrentThread()` if we see it. That will fix this specific issue but other phases might still do something similar. We can also mark all allocations as "have side-effects" in the presense of `GetAllocatedBytesForCurrentThread` but that probably a bad idea (observing impacts results).

**Comment by jkotas** [comment]
> To basically always spill GC.GetAllocatedBytesForCurrentThread() if we see it. That will fix this specific issue but other phases might still do something similar.

Right. I do not think special casing GetAllocatedBytesForCurrentThread in the JIT makes sense.

**Comment by AndyAyersMS** [comment]
> The JIT is reordering the allocation of the `Int32` box to be in front of `GetAllocatedBytesForCurrentThread` call as a side-effect of various optimizations. @dotnet/jit-contrib This reordering does not seem to be profitable. Do we want to avoid it?

As Egor mentioned, the benefit (more apparent with larger value types that are returned by ref) is that the code does not need to copy the value from the call's return buffer into the box payload; instead the call writes the value into the box directly.

**Comment by madelson** [comment]
Would adding something like a memory barrier inside the implementation of GetAllocatedBytesForCurrentThread be sufficient to prevent this reordering without any JIT changes?

My (possibly mistaken) assumption is that this methods only real purpose is measurements like what I’ve shown; it would be nice if those behaved as expected.

**Comment by jkotas** [comment]
> Would adding something like a memory barrier inside the implementation of GetAllocatedBytesForCurrentThread be sufficient to prevent this reordering without any JIT changes?

It would not help.

To avoid unexpected interferences between the measured code and the surrounding code, you can wrap the measured code in no-inline method. Wrappers like this are common for perf microbenchmark. For example, [Benchmark.NET](https://github.com/dotnet/BenchmarkDotNet) does something similar.
```csharp
[MethodImplAttribute(MethodImplOptions.NoInlining)]
static int Measure()
{
	var before = GC.GetAllocatedBytesForCurrentThread();
	var delta = GC.GetAllocatedBytesForCurrentThread() - before;
}

Console.WriteLine((object)Measure());
```
